cpp/ql/src/Security/CWE/CWE-611/Xerces.qll

```diff
@@ -67,8 +67,7 @@ class XercesDomParserLibrary extends XmlLibrary {
     // constructor.
     exists(CallInstruction call |
       call.getStaticCallTarget() = any(XercesDomParserClass c).getAConstructor() and
-      node.asInstruction().(WriteSideEffectInstruction).getDestinationAddress() =
-        call.getThisArgument() and
+      node.asInstruction().(StoreInstruction).getSourceValue() = call.getThisArgument() and
```

Comment:
Well spotted! The original (i.e., the `node.asInstruction().(WriteSideEffectInstruction)`) definitely wouldn't work since write side effects don't have [instruction nodes](https://github.com/github/codeql/blob/mathiasvp/replace-ast-with-ir-use-usedataflow/cpp/ql/lib/experimental/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll#L27).

I think this fix is okay. A couple of points are worth mentioning, though:
1. It slightly transforms what is being tracked: the original code tracks the allocation (by picking the write side effect as the source), but the new code tracks the pointer pointing to the new allocation.
2. Sicne the code is specifically looking for `StoreInstruction`s, it works on instances such as:
```cpp
p = new XercesDOMParser();
```
but I don't think it works on things like:
```cpp
void useDomParser(XercesDOMParser* p) { ... }
...
useDomParser(new XercesDOMParser());
```
In order to fix this, one could use a more literal translation of the old code such as:
```ql
override predicate configurationSource(DataFlow::Node node, string flowstate) {
  // source is the write on `this` of a call to the `XercesDOMParser`
  // constructor.
  exists(Call call |
    call.getTarget() = any(XercesDomParserClass c).getAConstructor() and
    node.asExpr() = call and
    encodeXercesFlowState(flowstate, 0, 1) // default configuration
  )
}
```
Because `node.asExpr()` maps a constructor call to the underlying `DataFlow::IndirectArgumentOutNode` (this is special-cased [here](https://github.com/github/codeql/blob/mathiasvp/replace-ast-with-ir-use-usedataflow/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L778)), which represents the value of `*pointer` after leaving the constructor, this will track the value of the value that's pointed to by the constructor. Then, changing `configurationSink` to use `asIndirectConvertedExpr` instead of `asConvertedExpr` should hopefully give the same flow.

Does that make sense?

Reply:
Hm! It seems to work locally for me 🤔. Did you also change the sink as I suggested here:
> Then, changing `configurationSink` to use `asIndirectConvertedExpr` instead of `asConvertedExpr` should hopefully give the same flow.
?

Reply:
Ah, I missed that bit.

Reply:
> I'm not sure if we were only supposed to capture cases where the allocation happened using new? I could imagine a stack allocation.

Indeed, I don't see why we shouldn't also support stack allocation.

Reply:
> Hm! It seems to work locally for me 🤔. Did you also change the sink as I suggested here:

That gives me too many results (things that are clearly marked as `GOOD`). Also changing the barrier gives me too few results.

Reply:
Oh, interesting. In that case, I think we should get this PR in since it's clearly better than what we currently have on the feature branch.

Reply:
I think I'll go for the solution with `NewExpr` though, that makes slightly more sense given the above. Also just stack allocating the parser variable, doesn't currently work either.